### PR TITLE
feat: Add support for orderBy and lockHint Spanner options

### PR DESCRIPTION
**Pull request by Hectorhammett**
Adds support for Spanner's OrderBy and LockHint options. 

b//381109733

### CODE REVIEW

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 1]
```diff
@@ -1352,7 +1352,9 @@ public async Task CanExecuteReadCommand()
             spannerClientMock.Received().StreamingRead(
                 Arg.Is<ReadRequest>(request => request.Columns.Equals(new RepeatedField<string> { "Col1", "Col2" })), Arg.Any<CallSettings>());
             spannerClientMock.Received().StreamingRead(
-                Arg.Is<ReadRequest>(request => request.KeySet.Equals(new V1.KeySet { Keys = {  new ListValue { Values =
+                Arg.Is<ReadRequest>(request => request.KeySet.Equals(new V1.KeySet
+                {
+                    Keys = {  new ListValue { Values =
```
Comment (Hectorhammett) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: I used the VScode format and using the recommended extension for `editorconfig` and this got changed but I don't think this is correct?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 2]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+        /// <summary>
+        /// Returns a clone of this ReadOptions with the given orderBy value.
+        /// </summary>
+        /// <param name="orderBy">
+        /// The value to use for the orderBy mechanism for the ReadOptions.
+        /// Only available for read-write transactions
+        /// Refer to the <see cref="Spanner.Data.OrderBy"/> enum to see the available value options.
+        /// </param>
+        /// <returns>A clone of this ReadOptions with the given ReadOptions value.</returns>
+        public ReadOptions WithOrderBy(OrderBy orderBy) =>
```
Comment (robertvoinescu-work) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Any reason the type isn't nullable i.e. should this be instead `WithOrderBy(OrderBy orderBy)` -> `WithOrderBy(OrderBy? orderBy)`? This would be consistent with the other nullable field `Limit? Limit`.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 3]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+
+        /// <summary>
+        /// Returns a clone of this ReadOptions with the given lockHint value.
+        /// </summary>
+        /// <param name="lockHint">
+        /// The value to use for the lockHint mechanism for the ReadOptions.
+        /// Refer to the <see cref="Spanner.Data.LockHint"/> enum to see the available value options.
+        /// </param>
+        /// <returns>A clone of this ReadOptions with the given lockHint value.</returns>
+        public ReadOptions WithLockHint(LockHint lockHint) =>
```
Comment (robertvoinescu-work) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Same comment here as below, should this be `WithLockHint(LockHint lockHint)` - > `WithLockHint(LocHint? lockHint)`?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs (hunk 2 above)
Comment (Hectorhammett) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Mostly because there is no null value for the ENUMs and I don't see a reason to set it null by yourself as realistically you would create a new readOptions for different needs.

If we agree that this needs to be changed though, I am happy to do so :)

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs [hunk 4]
```diff
@@ -718,5 +719,59 @@ public async Task ReadInterval_Array()
             var interval = Assert.Single(intervalList);
             Assert.Equal(expected, interval);
         }
+
+        [Fact]
+        public async Task ReadTakesLockHintOption()
```
Comment (purva9413) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: Can we combine the LockHint and OrderBy test as one integration test targeting a generic 'ReadOptionsAllowedByBackend' (or similar) test? My thought is to just have this one test for checking ReadOptions which are accepted by the Spanner backend. We do not need to test if they actually 'function' as they are supposed to, but just check they are set and formatted correctly for Spanner backend to accept

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs [hunk 5]
```diff
@@ -718,5 +719,59 @@ public async Task ReadInterval_Array()
...
+            using var reader = await cmd.ExecuteReaderAsync();
+            var count = 0;
+            while (await reader.ReadAsync())
+            {
+                count++;
+            }
+            Assert.Equal(_fixture.RowCount, count);
+        }
+
+        [Fact]
```
Comment (purva9413) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: Can we cover this in a unit test? Not sure we need a integration test just to see read only transactions do not accept these options.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 6]
```diff
@@ -76,7 +76,7 @@ public void CloneQuery()
         {
             var connection = new SpannerConnection("Data Source=projects/p/instances/i/databases/d");
             var command = connection.CreateSelectCommand("SELECT * FROM FOO");
-            var command2 = (SpannerCommand)command.Clone();
+            var command2 = (SpannerCommand) command.Clone();
```
Comment (purva9413) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: Nit: Please revert these spacing and lint changes

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 7]
```diff
@@ -1468,6 +1477,52 @@ public async Task CanExecuteReadCommandWithLimit()
...
+            Assert.True(reader.HasRows);
+
+            spannerClientMock.Received(1).StreamingRead(
+                Arg.Is<ReadRequest>(request => request.LockHint == ReadRequest.Types.LockHint.Exclusive),
+                Arg.Any<CallSettings>()
+            );
+        }
+
+        [Fact]
+        public async Task ExecuteReaderReadWithOrderBy()
```
Comment (purva9413) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: Can we make this test a `Theory` with `InlineData` and test all the viable values for the OrderBy enum are converted correctly to the ones expected by Spanner?

Same for the LockHint unit test as well.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 8]
```diff
@@ -40,6 +42,16 @@ public sealed class ReadOptions
         /// </summary>
         public long? Limit { get; }
 
+        /// <summary>
+        /// A lock hint mechanism for reads done within a transaction.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Are there links to Spanner documentation that we can include here and in the other one? Or if you have some of that on the enums themselves, maybe use `See<see ref="LockHint" /> for details.`

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs (hunk 2 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: +1 to what Robert said. The properties are nullable, as they should be, and making the parameters nullable is just consistent. And a user that wants exactly these options but with one of these unset shouldn't need to create new ones and figure out which of the other values to use there.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs (hunk 3 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: +1 to Robert's comment

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 9]
```diff
@@ -104,4 +141,97 @@ private static string CheckNotEmpty(string val, string paramName) =>
                 ? throw new ArgumentException("An empty string was provided, but is not valid", paramName)
                 : val;
     }
+
+    /// <summary>
+    /// The options abailable for the LockHint mechanism on a ReadRequest.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
    /// The options available for the LockHint mechanism on a ReadRequest.
```

But users at this level know nothing about a ReadRequest, so let's not mention that here.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 10]
```diff
@@ -104,4 +141,97 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// access the data until your transaction is ready to commit. When your
+        /// transaction is committing and writes are being applied, the transaction
+        /// attempts to upgrade to an exclusive lock for any data you are writing.
+        /// For more information about locks, see [Lock
+        /// modes](https://cloud.google.com/spanner/docs/introspection/lock-statistics#explain-lock-modes).
+        /// </summary>
+        Shared = 1,
+
+        /// <summary>
+        /// Acquire exclusive locks.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs:165]: In general these texts are too big to be in the summary. I think we can add here a short description and include a link to Spanner documentation about these options. Or, if we don't have that, then use `See <see cref="V1.ReadRequest.Types.LockHint">`.

And the same applies for OrderBy.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 11]
```diff
@@ -104,4 +141,97 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// **Note:** Request exclusive locks judiciously because they block others
+        /// from reading that data for the entire transaction, rather than just when
+        /// the writes are being performed. Unless you observe high write contention,
+        /// you should use the default of shared read locks so you don't prematurely
+        /// block other clients from reading the data that you're writing to.
+        /// </summary>
+        Exclusive = 2
+    }
+
+    internal class LockHintConverter
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
    internal static class LockHintConverter
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 12]
```diff
@@ -104,4 +141,97 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// </summary>
+        PrimaryKey = 1,
+
+        /// <summary>
+        /// Read rows are returned in any order.
+        /// </summary>
+        NoOrder = 2,
+    }
+
+    internal class OrderByConverter
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
    internal static class OrderByConverter
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 13]
```diff
@@ -104,4 +141,97 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+
+    /// <summary>
+    /// The options available for the OrderBy option for a ReadRequest.
+    /// </summary>
+    public enum OrderBy
+    {
+        /// <summary>
+        /// Default value.
+        ///
+        /// `ORDER_BY_UNSPECIFIED` is equivalent to `ORDER_BY_PRIMARY_KEY`.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: In particular, these values here match the API, but not the library values. This is confusing.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs (hunk 6 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: Or add them all in a separate commit 'chore: Lint changes'.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs (hunk 1 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: I think it's fine, but we do not usually include formatting and style unrelated changes on a feature commit, just to make it easier to review and reason about the change itself.

Either rever the linter changes or add them in separate, specific commit.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs (hunk 7 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: +1

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs (hunk 5 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: We are not validating, and that's fine, we avoid these types of validation client side. But then we also don't need this test, which only tests for server behaviour, no client side behaviour is being tested here, so we don't need it.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs (hunk 4 above)
Comment (amanda-tarafa) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: Maybe not? We already have a specific ReadWithLimit, and at some point some of these options may not work together (server side), so let's keep a single test per option.

@Hectorhammett but maybe change the name to ReadWithLockHint, ReadWithOrderBy for consistency.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs [hunk 14]
```diff
@@ -14,6 +14,7 @@
 
 using Google.Cloud.Spanner.Data.CommonTesting;
 using Google.Cloud.Spanner.V1;
+using Google.Rpc;
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: What do we need this for?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 15]
```diff
@@ -13,14 +13,16 @@
 // limitations under the License.
 
 using Google.Api.Gax;
+using Google.Apis.Util;
 using System;
 using System.Collections.Generic;
+using System.Resources;
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Do we need these new usings?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs (hunk 6 above)
Comment (Hectorhammett) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs]: Yeah, I wonder why is the lint doing this for me. I am using the recommended VScode Microsoft pluggin? But it seems to not work correctly for me. Will have to take a deeper look.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs (hunk 14 above)
Comment (Hectorhammett) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: I think VScode auto imported something while hitting an autocomplete value. This is a mistake and has been taken off.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs (hunk 15 above)
Comment (Hectorhammett) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]: Same issue as above, sometimes I use the wrong value on the autocomplete and it auto imports what I don't need. Fixed.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 16]
```diff
@@ -40,6 +40,18 @@ public sealed class ReadOptions
         /// </summary>
         public long? Limit { get; }
 
+        /// <summary>
+        /// A lock hint mechanism for reads done within a transaction.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// A lock hint for reads within a transaction.
        /// May be null, in which case appropiate defaults will be used.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 17]
```diff
@@ -40,6 +40,18 @@ public sealed class ReadOptions
...
+        /// See <see cref="LockHint"/> for details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// See <see cref="Spanner.Data.LockHint"/> for details.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 18]
```diff
@@ -40,6 +40,18 @@ public sealed class ReadOptions
...
+        /// </summary>
+        public LockHint? LockHint { get; }
+
+        /// <summary>
+        /// An option to control the order in which rows are returned from a read.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// The order in which rows are returned from a read.
        /// May be null, in which case appropiate defaults will be used.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 19]
```diff
@@ -40,6 +40,18 @@ public sealed class ReadOptions
...
+        /// See <see cref="OrderBy"/> for details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// See <see cref="Spanner.Data.OrderBy"/> for details.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 20]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
         /// </param>
         /// <returns>A clone of this ReadOptions with the given index name.</returns>
         public ReadOptions WithIndexName(string indexName) =>
-            new ReadOptions(Columns, CheckNotEmpty(indexName, nameof(indexName)), Limit);
+            new ReadOptions(Columns, CheckNotEmpty(indexName, nameof(indexName)), Limit, LockHint, OrderBy);
+
+        /// <summary>
+        /// Returns a clone of this ReadOptions with the given lockHint value.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Returns a clone of this ReadOptions with the given lock hint.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 21]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+        /// </summary>
+        /// <param name="lockHint">
+        /// The value to use for the lockHint mechanism for the ReadOptions.
+        /// Refer to the <see cref="Spanner.Data.LockHint"/> enum to see the available value options.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// The new lock hint value. May be null in which case appropiate defaults will be used.
        /// See <see cref="Spanner.Data.LockHint"/> for available options.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 22]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+        /// </param>
+        /// <returns>A clone of this ReadOptions with the given lockHint value.</returns>
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// <returns>A clone of this ReadOptions with the given lock hint.</returns>
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 23]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+        public ReadOptions WithLockHint(LockHint? lockHint) =>
+            new ReadOptions(Columns, IndexName, Limit, lockHint, OrderBy);
+
+        /// <summary>
+        /// Returns a clone of this ReadOptions with the given orderBy value.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Returns a clone of this ReadOptions with the given order by value.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 24]
```diff
@@ -92,7 +106,30 @@ public ReadOptions WithLimit(long? limit) =>
...
+        /// </summary>
+        /// <param name="orderBy">
+        /// The value to use for the orderBy mechanism for the ReadOptions.
+        /// Only available for read-write transactions
+        /// Refer to the <see cref="Spanner.Data.OrderBy"/> enum to see the available value options.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// The order on which rows are returned from a read. May be null, in which case appropiate defaults will be used.
        /// Only read-write transactions accept non-null values.
        /// See <see cref="Spanner.Data.OrderBy"/> for available options.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 25]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
                 ? throw new ArgumentException("An empty string was provided, but is not valid", paramName)
                 : val;
     }
+
+    /// <summary>
+    /// The options available for the LockHint value on the ReadOptions.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
    /// Lock hint options for reads within a transaction.
    /// See <see cref="Spanner.V1.ReadRequest.Types.LockHint"/> for details.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 26]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+    public enum LockHint
+    {
+        /// <summary>
+        /// Default value.
+        ///
+        /// `Unspecified` is equivalent to `Shared`.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.LockHint
+        /// for further details.
+        /// </summary>
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs:156]:
```suggestion
        /// <summary>
        /// Default value. Equivalent to <see cref="Shared" />.
        /// </summary>
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 27]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        Unspecified = 0,
+
+        /// <summary>
+        /// Acquire shared locks.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.LockHint
+        /// for further details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Acquire shared locks.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 28]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// </summary>
+        Shared = 1,
+
+        /// <summary>
+        /// Acquire exclusive locks.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.LockHint
+        /// for further details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Acquire exclusive locks.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 29]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        Exclusive = 2
+    }
+
+    internal static class LockHintConverter
+    {
+        internal static V1.ReadRequest.Types.LockHint ToProto(LockHint lockHint) => (V1.ReadRequest.Types.LockHint) (int) lockHint;
+    }
+
+    /// <summary>
+    /// The options available for the OrderBy option for the ReadOptions.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
    /// Options for the order in which rows are returned from a read.
    /// See <see cref="Spanner.V1.ReadRequest.Types.OrderBy"/> for details.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 30]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+    /// </summary>
+    public enum OrderBy
+    {
+        /// <summary>
+        /// Default value.
+        ///
+        /// `Unspecified` is equivalent to `PrimaryKey`.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.OrderBy
+        /// for further details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Default value. Equivalent to <see cref="PrimaryKey" />.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 31]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// </summary>
+        Unspecified = 0,
+
+        /// <summary>
+        /// Read rows are returned in primary key order.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.OrderBy
+        /// for further details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Read rows are returned in primary key order.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs [hunk 32]
```diff
@@ -104,4 +141,78 @@ private static string CheckNotEmpty(string val, string paramName) =>
...
+        /// </summary>
+        PrimaryKey = 1,
+
+        /// <summary>
+        /// Read rows are returned in any order.
+        ///
+        /// See https://cloud.google.com/dotnet/docs/reference/Google.Cloud.Spanner.V1/latest/Google.Cloud.Spanner.V1.ReadRequest.Types.OrderBy
+        /// for further details.
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.ReadOptions.cs]:
```suggestion
        /// Read rows are returned in any order.
```

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs [hunk 33]
```diff
@@ -1468,6 +1468,58 @@ public async Task CanExecuteReadCommandWithLimit()
                 Arg.Any<CallSettings>());
         }
 
+        [Theory]
+        [InlineData(LockHint.Unspecified, ReadRequest.Types.LockHint.Unspecified)]
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.Tests/SpannerCommandTests.cs:1472]: Add one test for null => Unspecified. You'll need to make the first parameter of the test nullable.

And do the same for order by.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs [hunk 34]
```diff
@@ -718,5 +718,59 @@ public async Task ReadInterval_Array()
...
+            using var connection = _fixture.GetConnection();
+
+            var readOptions = ReadOptions.FromColumns("key", "StringValue")
+                .WithLockHint(LockHint.Exclusive);
+            using var cmd = connection.CreateReadCommand(
+                _fixture.TableName, readOptions, KeySet.All);
+            using var reader = await cmd.ExecuteReaderAsync();
+            var e = await Assert.ThrowsAsync<SpannerException>(() => reader.ReadAsync());
+            Assert.Equal(ErrorCode.InvalidArgument, e.ErrorCode);
+        }
```
Comment (amanda-tarafa) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.IntegrationTests/ReadTests.cs]: Let's remove this test. It's only testing server side behaviour as we are not validating client side (which is right)

**Review by amanda-tarafa — CHANGES_REQUESTED** [review]
Just documentation and test improvements. Otherwise, LGTM, thanks!
